File: app/src/main/java/com/ssy281/liveoverflow/overlay/OverlayService.java
package com.ssy281.liveoverflow.overlay;

import android.app.Notification;
import android.app.NotificationChannel;
import android.app.NotificationManager;
import android.app.PendingIntent;
import android.app.Service;
import android.app.usage.UsageStats;
import android.app.usage.UsageStatsManager;
import android.content.Context;
import android.content.Intent;
import android.graphics.Color;
import android.graphics.PixelFormat;
import android.os.Build;
import android.os.Handler;
import android.os.IBinder;
import android.os.Looper;
import android.view.Gravity;
import android.view.MotionEvent;
import android.view.View;
import android.view.WindowManager;
import android.view.animation.AlphaAnimation;
import android.view.animation.Animation;
import android.widget.FrameLayout;
import android.widget.ImageView;
import android.widget.TextView;
import android.widget.Toast;

import androidx.core.app.NotificationCompat;

import com.ssy281.liveoverflow.MainActivity;
import com.ssy281.liveoverflow.R;

import java.util.List;
import java.util.Random;

public class OverlayService extends Service {
    private WindowManager windowManager;
    private View overlayView;
    private WindowManager.LayoutParams params;
    private TextView bubbleView;
    private WindowManager.LayoutParams bubbleParams;

    private float initialX, initialY;
    private float initialTouchX, initialTouchY;
    private long downTime;
    private Handler handler;
    private String lastPkg = "";
    private String lastReaction = "";
    private long lastReactionTime = 0;

    private final Handler bubbleHandler = new Handler();
    private final Handler appCheckHandler = new Handler();
    private final Handler idleHandler = new Handler();
    private final Handler drinkReminderHandler = new Handler();

    private static final String[][] PHRASES = {
            {"戳我干嘛！", "嗯？", "诶嘿~", "在呢在呢"},
            {"许星阔！怎么啦？", "宝宝你戳到我了~", "嘶...好痒"},
            {"别揉啦，发型乱了！", "唔...轻点嘛", "抱抱就抱抱！", "呜哇，晕了晕了"},
    };

    private static final String[] IDLE_PHRASES = {
            "好无聊啊…", "宝宝在干嘛呢？", "想喝奶茶~", 
            "趴好了，等你来戳", "盯——", "喵..."
    };

    private final Random random = new Random();

    @Override
    public void onCreate() {
        super.onCreate();
        windowManager = (WindowManager) getSystemService(WINDOW_SERVICE);
        handler = new Handler(Looper.getMainLooper());
        createNotificationChannel();
        startForeground(1, buildNotification());
        createOverlayView();
        createBubbleView();
        startAppDetection();
        startIdleBubble();
        startDrinkReminder();
        // 检查使用情况访问权限
        checkUsageStatsPermission();
    }
    
    private void checkUsageStatsPermission() {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP_MR1) {
            UsageStatsManager usm = (UsageStatsManager) getSystemService(Context.USAGE_STATS_SERVICE);
            long now = System.currentTimeMillis();
            List<UsageStats> stats = usm.queryUsageStats(UsageStatsManager.INTERVAL_DAILY, now - 1000, now);
            if (stats == null || stats.isEmpty()) {
                showBubble("给我「使用情况访问」权限，就能看到切应用啦~");
            }
        }
    }

    private void createNotificationChannel() {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            NotificationChannel channel = new NotificationChannel(
                    "overlay_channel",
                    "许星阔桌宠",
                    NotificationManager.IMPORTANCE_LOW
            );
            channel.setDescription("桌宠悬浮窗运行中");
            NotificationManager manager = getSystemService(NotificationManager.class);
            if (manager != null) {
                manager.createNotificationChannel(channel);
            }
        }
    }

    private Notification buildNotification() {
        Intent intent = new Intent(this, MainActivity.class);
        PendingIntent pendingIntent = PendingIntent.getActivity(
                this, 0, intent, PendingIntent.FLAG_IMMUTABLE
        );

        return new NotificationCompat.Builder(this, "overlay_channel")
                .setContentTitle("许星阔桌宠")
                .setContentText("正在运行中…")
                .setSmallIcon(android.R.drawable.ic_dialog_info)
                .setContentIntent(pendingIntent)
                .setOngoing(true)
                .build();
    }

    private void createOverlayView() {
        overlayView = new ImageView(this);
        ((ImageView) overlayView).setImageResource(R.mipmap.ic_launcher);

        params = new WindowManager.LayoutParams(
                WindowManager.LayoutParams.WRAP_CONTENT,
                WindowManager.LayoutParams.WRAP_CONTENT,
                Build.VERSION.SDK_INT >= Build.VERSION_CODES.O
                        ? WindowManager.LayoutParams.TYPE_APPLICATION_OVERLAY
                        : WindowManager.LayoutParams.TYPE_PHONE,
                WindowManager.LayoutParams.FLAG_NOT_FOCUSABLE,
                PixelFormat.TRANSLUCENT
        );
        params.gravity = Gravity.TOP | Gravity.START;
        params.x = 100;
        params.y = 200;

        windowManager.addView(overlayView, params);

        overlayView.setOnTouchListener(new View.OnTouchListener() {
            @Override
            public boolean onTouch(View v, MotionEvent event) {
                switch (event.getAction()) {
                    case MotionEvent.ACTION_DOWN:
                        initialX = params.x;
                        initialY = params.y;
                        initialTouchX = event.getRawX();
                        initialTouchY = event.getRawY();
                        downTime = System.currentTimeMillis();
                        return true;
                    case MotionEvent.ACTION_MOVE:
                        params.x = (int) (initialX + event.getRawX() - initialTouchX);
                        params.y = (int) (initialY + event.getRawY() - initialTouchY);
                        windowManager.updateViewLayout(overlayView, params);
                        return true;
                    case MotionEvent.ACTION_UP:
                        long duration = System.currentTimeMillis() - downTime;
                        float dx = event.getRawX() - initialTouchX;
                        float dy = event.getRawY() - initialTouchY;
                        if (duration > 500) {
                            // 长按：按住超过半秒直接触发第三组撒娇台词，不受位移影响
                            onPetClick(3);
                        } else if (Math.abs(dx) < 15 && Math.abs(dy) < 15) {
                            // 短按：随机台词
                            onPetClick(random.nextInt(2) == 0 ? 1 : 2);
                        }
                        return true;
                }
                return false;
            }
        });
    }

    private void onPetClick(int type) {
        String[] pool = PHRASES[Math.min(type - 1, PHRASES.length - 1)];
        String msg = pool[random.nextInt(pool.length)];
        showBubble(msg);
        Toast.makeText(OverlayService.this, "许星阔在此~", Toast.LENGTH_SHORT).show();
    }

    private void createBubbleView() {
        bubbleView = new TextView(this);
        bubbleView.setPadding(24, 14, 24, 14);
        bubbleView.setTextColor(Color.WHITE);
        bubbleView.setBackgroundColor(0xCCFF6B6B);
        bubbleView.setTextSize(14);
        bubbleView.setGravity(Gravity.CENTER);
        bubbleView.setVisibility(View.GONE);
        bubbleView.setElevation(16f);

        bubbleParams = new WindowManager.LayoutParams(
                WindowManager.LayoutParams.WRAP_CONTENT,
                WindowManager.LayoutParams.WRAP_CONTENT,
                Build.VERSION.SDK_INT >= Build.VERSION_CODES.O
                        ? WindowManager.LayoutParams.TYPE_APPLICATION_OVERLAY
                        : WindowManager.LayoutParams.TYPE_PHONE,
                WindowManager.LayoutParams.FLAG_NOT_FOCUSABLE | WindowManager.LayoutParams.FLAG_NOT_TOUCHABLE,
                PixelFormat.TRANSLUCENT
        );
        bubbleParams.gravity = Gravity.TOP | Gravity.START;
        bubbleParams.x = params.x;
        bubbleParams.y = params.y - 160; // 在悬浮窗上方

        windowManager.addView(bubbleView, bubbleParams);
    }

    private void showBubble(String text) {
        bubbleHandler.removeCallbacksAndMessages(null);
        bubbleView.setText(text);
        bubbleView.setVisibility(View.VISIBLE);
        // 位置跟随悬浮窗
        bubbleParams.x = params.x;
        bubbleParams.y = params.y - 160;
        try {
            windowManager.updateViewLayout(bubbleView, bubbleParams);
        } catch (Exception ignored) {}

        // 淡入动画
        AlphaAnimation fadeIn = new AlphaAnimation(0f, 1f);
        fadeIn.setDuration(200);
        bubbleView.startAnimation(fadeIn);

        // 4秒后隐藏
        bubbleHandler.postDelayed(new Runnable() {
            @Override
            public void run() {
                AlphaAnimation fadeOut = new AlphaAnimation(1f, 0f);
                fadeOut.setDuration(300);
                fadeOut.setAnimationListener(new Animation.AnimationListener() {
                    @Override
                    public void onAnimationEnd(Animation animation) {
                        bubbleView.setVisibility(View.GONE);
                    }
                    @Override public void onAnimationStart(Animation animation) {}
                    @Override public void onAnimationRepeat(Animation animation) {}
                });
                bubbleView.startAnimation(fadeOut);
            }
        }, 4000);
    }

    private void startAppDetection() {
        appCheckHandler.postDelayed(new Runnable() {
            @Override
            public void run() {
                checkForegroundApp();
                appCheckHandler.postDelayed(this, 1200);
            }
        }, 1200);
    }

    private void checkForegroundApp() {
        String pkg = getForegroundPackage();
        // 如果始终返回自身包名，说明未授予"使用情况访问权限"
        if (pkg != null && pkg.equals(getPackageName())) {
            return; // 权限未授予时静默跳过，不重复提示
        }
        if (pkg != null && !pkg.equals(lastPkg)) {
            String reaction;
            if (pkg.contains("taobao") || pkg.contains("jingdong")) {
                reaction = "买东西？我帮你挑！";
            } else if (pkg.contains("bilibili") || pkg.contains("aweme")) {
                reaction = "在刷视频呢？";
            } else if (pkg.contains("weixin") || pkg.contains("tencent.mm") || pkg.contains("qq")) {
                reaction = "聊天呢？我也在~";
            } else if (pkg.contains("com.android")) {
                reaction = "";
            } else {
                reaction = "换软件啦？";
            }
            if (!reaction.isEmpty() && !reaction.equals(lastReaction) && 
                    System.currentTimeMillis() - lastReactionTime > 10000) {
                showBubble(reaction);
                lastReaction = reaction;
                lastReactionTime = System.currentTimeMillis();
            }
            lastPkg = pkg;
        }
    }

    private String getForegroundPackage() {
        try {
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP_MR1) {
                UsageStatsManager usm = (UsageStatsManager) getSystemService(Context.USAGE_STATS_SERVICE);
                long time = System.currentTimeMillis();
                List<UsageStats> stats = usm.queryUsageStats(
                        UsageStatsManager.INTERVAL_DAILY, time - 8000, time);
                if (stats != null && !stats.isEmpty()) {
                    UsageStats recent = null;
                    for (UsageStats s : stats) {
                        if (recent == null || s.getLastTimeUsed() > recent.getLastTimeUsed()) {
                            recent = s;
                        }
                    }
                    if (recent != null) return recent.getPackageName();
                }
            }
        } catch (Exception e) {
            return getPackageName();
        }
        return getPackageName();
    }

    private void startIdleBubble() {
        idleHandler.postDelayed(new Runnable() {
            @Override
            public void run() {
                if (random.nextInt(100) < 60) {
                    String msg = IDLE_PHRASES[random.nextInt(IDLE_PHRASES.length)];
                    showBubble(msg);
                }
                idleHandler.postDelayed(this, 30000 + random.nextInt(60000));
            }
        }, 60000);
    }

    private void startDrinkReminder() {
        drinkReminderHandler.postDelayed(new Runnable() {
            @Override
            public void run() {
                showBubble("该喝水啦！");
                drinkReminderHandler.postDelayed(this, 45 * 60 * 1000);
            }
        }, 45 * 60 * 1000);
    }

    @Override
    public void onDestroy() {
        super.onDestroy();
        appCheckHandler.removeCallbacksAndMessages(null);
        idleHandler.removeCallbacksAndMessages(null);
        drinkReminderHandler.removeCallbacksAndMessages(null);
        bubbleHandler.removeCallbacksAndMessages(null);
        if (overlayView != null && windowManager != null) {
            windowManager.removeView(overlayView);
        }
        if (bubbleView != null && windowManager != null) {
            windowManager.removeView(bubbleView);
        }
    }

    @Override
    public IBinder onBind(Intent intent) {
        return null;
    }
}

File: app/src/main/java/com/ssy281/liveoverflow/overlay/OverlayService.kt
package com.ssy281.liveoverflow.overlay

import android.animation.ObjectAnimator
import android.app.NotificationChannel
import android.app.NotificationManager
import android.app.PendingIntent
import android.app.Service
import android.content.Context
import android.content.Intent
import android.graphics.Color
import android.graphics.PixelFormat
import android.os.Build
import android.os.IBinder
import android.view.Gravity
import android.view.MotionEvent
import android.view.View
import android.view.WindowManager
import android.widget.ImageView
import androidx.core.app.NotificationCompat
import com.ssy281.liveoverflow.MainActivity
import com.ssy281.liveoverflow.R
import kotlinx.coroutines.*

class OverlayService : Service() {
    private lateinit var wm: WindowManager
    private lateinit var overlayView: ImageView
    private var gestureEngine: GestureEngine? = null
    private var appDetector: AppDetector? = null
    private var heatSystem: HeatSystem? = null
    private var moodEngine: MoodEngine? = null
    private var bubbleManager: BubbleManager? = null
    private val scope = CoroutineScope(Dispatchers.Main + SupervisorJob())
    private var heatOverlay: View? = null

    private val phrases = mapOf(
        "daily" to listOf("今天天气真好呀~", "想喝奶茶了...", "看到你就开心！"),
        "clingy" to listOf("你别走嘛~", "再陪我一会儿好不好？", "我一个人好无聊哦..."),
        "chaos" to listOf("啊啊啊好多事！", "脑袋要爆炸了！", "今天是什么日子来着？"),
        "night" to listOf("这么晚还不睡...", "晚安啦宝宝~", "偷偷跟你说，我有点困了...")
    )

    private val appReactions = mapOf(
        "com.taobao.taobao" to "又买东西！让我看看~",
        "com.jingdong.app.mall" to "京东？又要剁手了是吧！",
        "com.ss.android.ugc.aweme" to "刷抖音不陪我...我吃醋了！",
        "com.tencent.wework" to "上班加油哦~",
        "com.alibaba.android.rimet" to "钉钉又响了，快去处理~",
        "com.tencent.mm" to "回微信呢？记得回我消息哦！"
    )

    override fun onCreate() {
        super.onCreate()
        wm = getSystemService(WINDOW_SERVICE) as WindowManager
        bubbleManager = BubbleManager(this)
        moodEngine = MoodEngine()

        gestureEngine = GestureEngine(
            onClick = { onPetClick(1) },
            onDoubleClick = { onPetClick(2) },
            onLongPress = { onPetClick(3) },
            onFling = { bubbleManager?.showBubble("呜哇——！", BubbleStyle.ANGRY) }
        )

        appDetector = AppDetector(this) { pkg ->
            moodEngine?.let { engine ->
                val mood = engine.getMood(pkg, heatSystem?.heat ?: 0)
                val reaction = appReactions[pkg] ?: getPhraseForMood(mood)
                val style = when {
                    pkg.contains("taobao") || pkg.contains("jingdong") -> BubbleStyle.JEALOUS
                    pkg.contains("aweme") -> BubbleStyle.HEART
                    else -> BubbleStyle.NORMAL
                }
                bubbleManager?.showBubble(reaction, style)
            }
        }

        heatSystem = HeatSystem(
            onHeatChanged = { updateHeatOverlay(it) },
            onHeatMax = { bubbleManager?.showBubble("啊啊啊要爆炸了！！", BubbleStyle.ANGRY) }
        )

        appDetector?.start(scope)
        heatSystem?.start(scope)

        startSelfTalk()
        startIdleCheck()
        startScheduledBehavior()
    }

    override fun onStartCommand(intent: Intent?, flags: Int, startId: Int): Int {
        createNotificationChannel()
        val pendingIntent = PendingIntent.getActivity(
            this, 0,
            Intent(this, MainActivity::class.java),
            PendingIntent.FLAG_UPDATE_CURRENT or PendingIntent.FLAG_IMMUTABLE
        )
        val notification = NotificationCompat.Builder(this, "overlay_channel")
            .setContentTitle("许星阔桌宠")
            .setContentText(getNotificationWhisper())
            .setSmallIcon(R.drawable.ic_notification)
            .setContentIntent(pendingIntent)
            .setOngoing(true)
            .build()
        startForeground(1, notification)
        startNotificationWhisperUpdater()
        createOverlayView()
        return START_STICKY
    }

    override fun onBind(intent: Intent?): IBinder? = null

    override fun onDestroy() {
        super.onDestroy()
        scope.cancel()
        appDetector?.stop()
        heatSystem?.stop()
        removeOverlayView()
    }

    private fun createNotificationChannel() {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            val channel = NotificationChannel(
                "overlay_channel", "许星阔桌宠",
                NotificationManager.IMPORTANCE_LOW
            ).apply { description = "桌宠前台服务通知" }
            val nm = getSystemService(NotificationManager::class.java)
            nm.createNotificationChannel(channel)
        }
    }

    private fun createOverlayView() {
        overlayView = ImageView(this).apply {
            setImageResource(R.mipmap.ic_launcher)
            adjustViewBounds = true
            scaleType = ImageView.ScaleType.FIT_CENTER
        }

        val params = WindowManager.LayoutParams(
            200, 200,
            WindowManager.LayoutParams.TYPE_APPLICATION_OVERLAY,
            WindowManager.LayoutParams.FLAG_NOT_FOCUSABLE,
            PixelFormat.TRANSLUCENT
        ).apply { gravity = Gravity.TOP or Gravity.START }

        wm.addView(overlayView, params)

        var initialX = 0
        var initialY = 0
        var touchX = 0f
        var touchY = 0f

        overlayView.setOnTouchListener { v, event ->
            gestureEngine?.onTouch(v, event)
            when (event.action) {
                MotionEvent.ACTION_DOWN -> {
                    initialX = params.x
                    initialY = params.y
                    touchX = event.rawX
                    touchY = event.rawY
                }
                MotionEvent.ACTION_MOVE -> {
                    params.x = initialX + (event.rawX - touchX).toInt()
                    params.y = initialY + (event.rawY - touchY).toInt()
                    wm.updateViewLayout(overlayView, params)
                }
            }
            true
        }
    }

    private fun removeOverlayView() {
        try { wm.removeView(overlayView) } catch (_: Exception) {}
        try { wm.removeView(heatOverlay) } catch (_: Exception) {}
    }

    private fun onPetClick(taps: Int) {
        heatSystem?.addHeat(taps * 5)
        moodEngine?.onInteract()
        val msg = when (taps) {
            1 -> listOf("嗯？", "诶嘿~", "戳我干嘛！", "别闹~").random()
            2 -> listOf("啊啊别戳了！", "痒死啦！", "讨厌！").random()
            3 -> "……你到底要戳几下啦！"
            else -> "呜哇哇哇！！"
        }
        bubbleManager?.showBubble(msg, BubbleStyle.HEART)
    }

    private fun updateHeatOverlay(heat: Int) {
        try { wm.removeView(heatOverlay) } catch (_: Exception) {}
        if (heat > 30) {
            heatOverlay = View(this).apply {
                setBackgroundColor(Color.argb((heat * 2.55 * 0.3).toInt(), 255, 0, 0))
            }
            val params = WindowManager.LayoutParams(
                WindowManager.LayoutParams.MATCH_PARENT,
                WindowManager.LayoutParams.MATCH_PARENT,
                WindowManager.LayoutParams.TYPE_APPLICATION_OVERLAY,
                WindowManager.LayoutParams.FLAG_NOT_FOCUSABLE or WindowManager.LayoutParams.FLAG_NOT_TOUCHABLE,
                PixelFormat.TRANSLUCENT
            )
            wm.addView(heatOverlay, params)
        }
    }

    private fun getPhraseForMood(mood: MoodEngine.Mood): String {
        val pool = when (mood) {
            MoodEngine.Mood.SLEEPY -> phrases["night"]!!
            MoodEngine.Mood.LONELY -> phrases["clingy"]!!
            MoodEngine.Mood.HAPPY -> phrases["daily"]!!
            else -> phrases["daily"]!!
        }
        return pool.random()
    }

    private fun startSelfTalk() {
        scope.launch {
            while (isActive) {
                delay((30_000..120_000).random().toLong())
                val mood = moodEngine?.getMood() ?: MoodEngine.Mood.HAPPY
                if (mood != MoodEngine.Mood.SLEEPY) {
                    bubbleManager?.showBubble(getPhraseForMood(mood), BubbleStyle.NORMAL)
                }
            }
        }
    }

    private fun startIdleCheck() {
        scope.launch {
            while (isActive) {
                delay(60_000)
                val mood = moodEngine?.getMood() ?: MoodEngine.Mood.HAPPY
                val idleMsg = when (mood) {
                    MoodEngine.Mood.LONELY -> "好无聊哦...你还不理我吗？"
                    MoodEngine.Mood.SHY -> "偷偷看你一眼..."
                    MoodEngine.Mood.SLEEPY -> "呼...呼..."
                    else -> null
                }
                idleMsg?.let { bubbleManager?.showBubble(it, BubbleStyle.WHISPER) }
            }
        }
    }

    private fun startScheduledBehavior() {
        scope.launch {
            while (isActive) {
                delay(1_200_000) // 20 min
                if (Math.random() < 0.3) {
                    bubbleManager?.showBubble("该喝水啦！", BubbleStyle.HEART)
                }
            }
        }
    }

    private val nightWhispers = listOf("夜深了...", "偷偷醒着陪你", "晚安世界")
    private val morningWhispers = listOf("早安！新的一天~", "今天也要开心哦", "阳光真好！")
    private val dayWhispers = listOf("在你身边真好", "正在守护你~", "有什么需要帮忙的吗？")

    private fun getNotificationWhisper(): String {
        val hour = java.util.Calendar.getInstance().get(java.util.Calendar.HOUR_OF_DAY)
        return when (hour) {
            in 0..5 -> nightWhispers.random()
            in 6..9 -> morningWhispers.random()
            else -> dayWhispers.random()
        }
    }

    private fun startNotificationWhisperUpdater() {
        scope.launch {
            while (isActive) {
                delay(3_600_000)
                val nm = getSystemService(NotificationManager::class.java)
                val notification = NotificationCompat.Builder(this@OverlayService, "overlay_channel")
                    .setContentTitle("许星阔桌宠")
                    .setContentText(getNotificationWhisper())
                    .setSmallIcon(R.drawable.ic_notification)
                    .setOngoing(true)
                    .build()
                nm.notify(1, notification)
            }
        }
    }
}
